File: backend/store/serializers.py
from rest_framework import serializers
from .models import Category, Product, Order, OrderItem, AnonymousUser, CartItem


class CategorySerializer(serializers.ModelSerializer):
    product_count = serializers.SerializerMethodField()

    class Meta:
        model = Category
        fields = ['id', 'name', 'slug', 'description', 'product_count', 'created_at']

    def get_product_count(self, obj):
        return obj.products.filter(is_active=True).count()


class ProductSerializer(serializers.ModelSerializer):
    category_name = serializers.CharField(source='category.name', read_only=True)
    image_display = serializers.SerializerMethodField()

    class Meta:
        model = Product
        fields = [
            'id', 'name', 'slug', 'description', 'price', 'category', 
            'category_name', 'image', 'image_url', 'image_display', 
            'stock', 'available_sizes', 'is_featured', 'is_active',
            'created_at', 'updated_at'
        ]

    def get_image_display(self, obj):
        # Prioritize uploaded image over URL
        if obj.image:
            request = self.context.get('request')
            if request:
                return request.build_absolute_uri(obj.image.url)
        return obj.image_url


class AnonymousUserSerializer(serializers.ModelSerializer):
    class Meta:
        model = AnonymousUser
        fields = ['id', 'fingerprint', 'email', 'first_name', 'last_name', 'phone', 'created_at']


class CartItemSerializer(serializers.ModelSerializer):
    product = ProductSerializer(read_only=True)
    product_id = serializers.IntegerField(write_only=True)

    class Meta:
        model = CartItem
        fields = ['id', 'product', 'product_id', 'quantity', 'size', 'created_at', 'updated_at']


class OrderItemSerializer(serializers.ModelSerializer):
    product_name = serializers.CharField(source='product.name', read_only=True)
    product_image = serializers.SerializerMethodField()

    class Meta:
        model = OrderItem
        fields = ['id', 'product', 'product_name', 'product_image', 'quantity', 'size', 'price', 'total_price']

    def get_product_image(self, obj):
        if obj.product.image:
            request = self.context.get('request')
            if request:
                return request.build_absolute_uri(obj.product.image.url)
        return obj.product.image_url


class OrderSerializer(serializers.ModelSerializer):
    items = OrderItemSerializer(many=True, read_only=True)
    order_items = serializers.ListField(write_only=True, required=False)

    class Meta:
        model = Order
        fields = [
            'id', 'order_number', 'email', 'first_name', 'last_name', 'phone',
            'delivery_address', 'delivery_city', 'delivery_state',  # Changed
            'delivery_postal_code', 'delivery_country',  # Changed
            'total_amount', 'status', 'notes', 'payment_method',
            'items', 'order_items',
            'created_at', 'updated_at'
        ]
        read_only_fields = ['order_number']

    def create(self, validated_data):
        order_items_data = validated_data.pop('order_items', [])
        order = Order.objects.create(**validated_data)
        
        for item_data in order_items_data:
            OrderItem.objects.create(order=order, **item_data)
        
        return order

File: backend-node/src/lib/serializers.js
import config from '../config.js';
import { fromId } from './ids.js';

/** DRF renders DecimalField as a string with the field's decimal_places. */
const decimal = (value) => (value === null || value === undefined ? null : Number(value).toFixed(2));

/** DRF renders DateTimeField as ISO-8601 with a trailing Z. */
const datetime = (value) => (value ? new Date(value).toISOString() : null);

/** Absolute URL for a stored media path, e.g. "products/dress.png". */
export function mediaUrl(req, relativePath) {
  if (!relativePath) return null;
  const base = `${req.protocol}://${req.get('host')}`;
  return `${base}${config.mediaUrl}${String(relativePath).replace(/^\/+/, '')}`;
}

export function serializeCategory(category, { productCount = null } = {}) {
  return {
    id: fromId(category.id),
    name: category.name,
    slug: category.slug,
    description: category.description,
    product_count: productCount ?? category._count?.products ?? 0,
    created_at: datetime(category.createdAt),
  };
}

export function serializeProduct(product, req) {
  const uploaded = product.image ? mediaUrl(req, product.image) : null;
  return {
    id: fromId(product.id),
    name: product.name,
    slug: product.slug,
    description: product.description,
    price: decimal(product.price),
    category: fromId(product.categoryId),
    category_name: product.category?.name ?? null,
    image: uploaded,
    image_url: product.imageUrl,
    // Uploaded file wins over the pasted URL, matching the Django serializer.
    image_display: uploaded ?? product.imageUrl,
    stock: product.stock,
    available_sizes: product.availableSizes ?? [],
    is_featured: product.isFeatured,
    is_active: product.isActive,
    created_at: datetime(product.createdAt),
    updated_at: datetime(product.updatedAt),
  };
}

export function serializeOrderItem(item, req) {
  const product = item.product;
  return {
    id: fromId(item.id),
    product: fromId(item.productId),
    product_name: product?.name ?? null,
    product_image: product?.image ? mediaUrl(req, product.image) : (product?.imageUrl ?? null),
    quantity: item.quantity,
    size: item.size,
    price: decimal(item.price),
    total_price: Number((Number(item.price) * item.quantity).toFixed(2)),
  };
}

export function serializeOrder(order, req) {
  return {
    id: fromId(order.id),
    order_number: order.orderNumber,
    email: order.email,
    first_name: order.firstName,
    last_name: order.lastName,
    phone: order.phone,
    delivery_address: order.deliveryAddress,
    delivery_city: order.deliveryCity,
    delivery_state: order.deliveryState,
    delivery_postal_code: order.deliveryPostalCode,
    delivery_country: order.deliveryCountry,
    total_amount: decimal(order.totalAmount),
    status: order.status,
    notes: order.notes,
    payment_method: order.paymentMethod,
    items: (order.items ?? []).map((item) => serializeOrderItem(item, req)),
    created_at: datetime(order.createdAt),
    updated_at: datetime(order.updatedAt),
  };
}

export function serializeCartItem(item, req) {
  return {
    id: fromId(item.id),
    product: item.product ? serializeProduct(item.product, req) : null,
    quantity: item.quantity,
    size: item.size,
    created_at: datetime(item.createdAt),
    updated_at: datetime(item.updatedAt),
  };
}

export function serializeAdminUser(admin) {
  return {
    id: admin.id,
    username: admin.username,
    email: admin.email,
    is_active: admin.isActive,
    created_at: datetime(admin.createdAt),
    last_login: datetime(admin.lastLogin),
  };
}
